### PR TITLE
feat: complete security ACL tools with update/delete and bulk operations

### LINKED ISSUE #82: Complete security ACL tools
## Problem

Current ACL tools focus on listing and creating, with update/delete and bulk operations missing.

- **Server ACL**: only list/create; no update/delete
- **File ACL**: only list/create; no update/delete
- **Command ACL**: list/create/update/delete implemented (complete)
- **Server ACL bulk**: unable to apply ACL changes across many servers at once

## Current state

| Feature | MCP tool | Status |
|---------|----------|--------|
| `list_command_acls` | `list_command_acls` | ✅ |
| `create_command_acl` | `create_command_acl` | ✅ |
| `update_command_acl` | `update_command_acl` | ✅ |
| `delete_command_acl` | `delete_command_acl` | ✅ |
| `list_server_acls` | `list_server_acls` | ✅ |
| `create_server_acl` | `create_server_acl` | ✅ |
| Server ACL update | — | ❌ |
| Server ACL delete | — | ❌ |
| Server ACL bulk add/remove | — | ❌ |
| `list_file_acls` | `list_file_acls` | ✅ |
| `create_file_acl` | `create_file_acl` | ✅ |
| File ACL update | — | ❌ |
| File ACL delete | — | ❌ |

## Target API endpoints

| Endpoint | Method | Description |
|----------|--------|-------------|
| `/api/security/server-acl/{id}/` | PUT/PATCH | Update server ACL |
| `/api/security/server-acl/{id}/` | DELETE | Delete server ACL |
| `/api/security/server-acl/bulk/` | POST | Bulk add/remove server ACLs |
| `/api/security/file-acl/{id}/` | PUT/PATCH | Update file ACL |
| `/api/security/file-acl/{id}/` | DELETE | Delete file ACL |

## MCP tools to implement

```python
# tools/security_tools.py extension

update_server_acl(acl_id, workspace, region, ...)
delete_server_acl(acl_id, workspace, region)
bulk_server_acl(workspace, region, action, acl_list)   # action: add | remove
update_file_acl(acl_id, workspace, region, ...)
delete_file_acl(acl_id, workspace, region)
```

## Acceptance criteria

- [ ] Add missing tools to `tools/security_tools.py`
- [ ] Register new tools in `server.py`
- [ ] Validate bulk ACL operations
- [ ] Add unit tests (`tests/test_security_tools.py`)

**Affected files**: `tools/security_tools.py`
**Related docs**: `docs/gap-analysis-2026-04-30.md` §5.3

### PR DESCRIPTION
**Pull request by jisung-02**
## Summary

Completes the security ACL toolset (closes #82). Adds update/delete tools for server and file ACLs plus bulk server ACL operations, then hardens the whole module based on iterative review against the server-side implementation.

### New tools
- `update_server_acl` / `delete_server_acl` / `bulk_server_acl`
- `update_file_acl` / `delete_file_acl`

### Breaking changes
- `create_command_acl` and `create_file_acl` parameter schemas changed: the previous `effect`/`command_pattern`/`users`/`groups`/`servers`/`priority` parameters never matched the upstream API (`CommandACLSerializer` accepts `token`/`service_token`/`command`/`username`/`groupname` only), so those calls failed server-side. The new schemas mirror the actual serializer contracts. There is no working old behavior to alias, so no deprecation shim is provided.

### Review fixes
- **Cross-type token rebind**: `update_server_acl` now explicitly nulls the opposing token field on PATCH—without this, rebinding an api-token-bound ACL to a service token violates the server's `exactly_one_token_type` check constraint and fails with a misleading `ACL_EXISTS` error
- Mutual exclusivity guards (`api_token_id` vs `service_token_id`) on all list/create/update/bulk functions
- `bulk_server_acl` guards: action whitelist, empty list, and 100-item cap (mirrors server-side `ServerACLBulk*Serializer` limits)
- Docstrings clarified: token rebind is supported by `update_server_acl` only; command/file ACL updates require delete-and-recreate

### Code quality
- Module constants for endpoints (`_API_*`), valid actions (`VALID_*`), and repeated error messages, following `webftp_tools`/`iam_tools` conventions
- Consistent validation order across functions (value-format checks before token checks)
- Test classes reordered to match source declaration order

### Endpoint contract note
Bulk remove uses `POST /api/security/server-acl/bulk/delete/`—verified against the server's `ServerACLViewSet.bulk_delete` action (`url_path='bulk/delete'`). Issue #82's description listing only `/bulk/` for both directions is imprecise.

## Test plan
- [x] 50 unit tests pass (`tests/test_security_tools.py`), covering payload construction, endpoint routing, and all validation error paths
- [ ] Verify cross-type token rebind (`update_server_acl` api token ↔ service token) against a dev workspace

Closes #82

### CODE REVIEW

File: tools/security_tools.py [hunk 1]
```diff
@@ -55,60 +83,57 @@ async def list_command_acls(
...
-    effect: str,
-    command_pattern: str,
-    users: list[str] | None = None,
-    groups: list[str] | None = None,
-    servers: list[str] | None = None,
-    description: str | None = None,
-    priority: int | None = None,
+    command: str,
+    api_token_id: str | None = None,
+    service_token_id: str | None = None,
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply tools/security_tools.py:90]: Intentional breaking change. The previous schema (`effect`/`command_pattern`/`users`/`groups`/`servers`/`priority`) never matched the upstream API—the server's `CommandACLSerializer` only accepts `token`/`service_token`/`command`/`username`/`groupname`, so those calls failed at the server with a missing-`command` validation error. There is no working old behavior to preserve via deprecated aliases. Documented as a breaking change in the PR description.

File: tools/security_tools.py [hunk 2]
```diff
@@ -265,62 +287,216 @@ async def list_server_acls(
...
+    token = kwargs.get('token')
+
+    body: dict[str, Any] = {'servers': server_ids}
+    if api_token_id is not None:
+        body['token'] = api_token_id
+    else:
+        body['service_token'] = service_token_id
+
+    endpoint = _API_SERVER_ACL_BULK if action == 'add' else _API_SERVER_ACL_BULK_DELETE
+
```
[the remark this answers is not in the document]
Comment (jisung-02) [reply tools/security_tools.py:480]: Confirmed against the server source: `ServerACLViewSet` defines `bulk_create` with `url_path='bulk'` and `bulk_delete` with `url_path='bulk/delete'` (security/api/views.py). The implementation is correct; issue #82's description listing only `/bulk/` was imprecise. Added an endpoint contract note to the PR description.

**Review by junho226 — COMMENTED** [review]
## PR Review: #113 — feat: complete security ACL tools with update/delete and bulk operations

**Author**: @jisung-02
**Branch**: `82-feat-complete-acl-tools` → `main`
**Stack**: Python (MCP server)
**Changes**: +1473 −126 across 3 files

---

### Summary

Completes the security ACL toolset (closes #82) by adding `update`/`delete` for server and file ACLs plus `bulk_server_acl`. Also rewrites `create_command_acl` and `create_file_acl` payload schemas to match the real upstream serializers (`CommandACLSerializer` / `FileACLSerializer`) — the previous schemas (`effect`/`command_pattern`/`users`/`groups`/`servers`/`priority`) never matched the server contract, so this is a corrective breaking change rather than a regression. Adds 50 unit tests, module constants for endpoints/error messages, and consistent mutual-exclusivity validation across all functions.

### Critical issues (0)

None.

### Warnings (1)

1. **`tools/security_tools.py` — `bulk_server_acl` is annotated `DESTRUCTIVE` even though `action='add'` is purely additive.** The annotation is fixed at decorator-time and can't reflect runtime behavior, so the add path inherits the strictest treatment from MCP clients (extra confirmation prompts, etc.). Consider either:
   - Splitting into `bulk_add_server_acl` (`ADDITIVE`) and `bulk_remove_server_acl` (`DESTRUCTIVE`), which also clarifies intent at tool-discovery time, **or**
   - Adding a one-line comment justifying why the conservative annotation is intentional (i.e., "remove path dominates").

   Not blocking, but worth a decision before this becomes load-bearing.

### Suggestions (4)

1. **`tools/security_tools.py` — bulk action error message hardcodes the valid set.**
   ```python
   if action not in VALID_BULK_ACTIONS:
       return error_response(f"Invalid action '{action}'. Must be 'add' or 'remove'.")
   ```
   You already built `_FILE_ACL_ACTIONS_STR` from `VALID_FILE_ACL_ACTIONS` for the same purpose. Apply the same pattern to `VALID_BULK_ACTIONS` so the message can't drift if the set changes later. Note `VALID_BULK_ACTIONS` is a `frozenset`, so ordering would be non-deterministic — use `sorted(...)` or switch the constant to a tuple for stable output.

2. **`tools/security_tools.py` — asymmetric defaulting of `username` / `groupname` between create and update.**
   - `create_*` sends `username=''` / `groupname=''` unconditionally (always present in payload).
   - `update_*` only includes them when explicitly set (`None` is treated as "don't touch").

   This is a defensible PATCH-vs-POST distinction, but worth a docstring note on the create path that "Empty = token owner only" applies on create and is *not* the way to clear an existing restriction on update (the user has to pass `username=''` explicitly there).

3. **`tests/test_security_tools.py` — missing boundary test for the 100-item bulk cap.** `test_bulk_over_100_server_ids_returns_error` exercises 101 (reject). Add a sibling test for exactly 100 (accept) so a future off-by-one in `_BULK_MAX_SERVERS` enforcement gets caught.

4. **PR test plan — `Verify cross-type token rebind ... against a dev workspace` is still unchecked.** The whole `update_server_acl` rebind path (nulling the opposing token field to satisfy `exactly_one_token_type`) is the most server-side-coupled change in this PR. The unit tests verify the request body shape, but not that the server actually accepts the `None` semantics. Worth confirming against a live workspace before merge, since a misreading of the server check constraint would surface as a confusing `ACL_EXISTS` error in production.

### Highlights

- **Excellent rationale capture** in the PR description — the breaking-change reasoning (server-side serializer never accepted the old shape) and the bulk endpoint contract note (`bulk/delete/` confirmed against `ServerACLViewSet`) preempt the exact questions a reviewer would ask. Both Copilot rounds already converged on these.
- **Module constants** for endpoints, valid action sets, and repeated error messages — clean factoring, no duplication.
- **Cross-type rebind comment** at `update_server_acl` ("the server enforces exactly-one-token-type, and PATCH keeps the old binding otherwise") is exactly the kind of WHY-comment that earns its keep. It explains a non-obvious server constraint that future readers would otherwise hit in production.
- **Bulk validation order** (action whitelist → empty list → 100-item cap → token mutual exclusivity) is consistent and matches the upstream serializer's order of checks.
- **Test coverage** is thorough across payload shape, endpoint routing, and every validation error path. Test classes mirror source declaration order — easy to navigate.

---

**Verdict**: COMMENTED — no blocking issues; the one warning and four suggestions are all incremental improvements. The unchecked manual verification of cross-type token rebind is the only thing I'd want closed before merge.

**Checklist**:
- [x] Code follows stack conventions (matches `webftp_tools`/`iam_tools` patterns per PR description)
- [x] Tests included and comprehensive (50 unit tests)
- [x] No security vulnerabilities identified
- [x] Documentation updated (`CLAUDE.md` reflects new tools)
- [ ] Cross-type token rebind verified against dev workspace (still pending per PR description)

---
🤖 Auto-review by junho-bot during junho226's leave. If this review looks off, please disregard — junho will follow up upon return.

**Comment by jisung-02** [comment]
@junho226 Addressed your review in 54848a7:

- **Warning (DESTRUCTIVE on `bulk_server_acl`)**: kept the single tool—splitting would churn the API again and the server exposes the two directions as one endpoint pair. Added the one-line comment justifying the conservative annotation ("annotations are fixed at decorator-time; the destructive 'remove' path dominates").
- **Suggestion 1**: `VALID_BULK_ACTIONS` is now a tuple (stable ordering) and the error message is derived via `_BULK_ACTIONS_STR`, same pattern as `_FILE_ACL_ACTIONS_STR`.
- **Suggestion 2**: documented the clear semantics on the **update** side (`update_command_acl`/`update_file_acl` docstrings: pass `''` explicitly to clear, omit to leave unchanged)—that's where the clearing action happens, so the note lands where users hit it.
- **Suggestion 3**: added `test_bulk_exactly_100_server_ids_succeeds` (51 tests passing).
- **Suggestion 4**: still open—the connected dev MCP server doesn't expose `update_server_acl` until this PR ships, so the cross-type rebind needs manual verification against a dev workspace before merge.

**Review by junho226 — DISMISSED** [review]
## Follow-up review: PR #113 — iteration on 54848a7

**Author**: @jisung-02
**Branch**: `82-feat-complete-acl-tools` → `main`
**Stack**: Python (MCP server)
**Re-review scope**: delta since prior review at 120c279 (3 file changes: +34 −6)

---

### Summary

All four suggestions from the prior review are addressed in 54848a7. Re-reviewed only the delta plus a sanity pass on the surrounding code.

### Resolution check

| Prior item | Resolution | Verdict |
|---|---|---|
| Warning: `bulk_server_acl` DESTRUCTIVE annotation | Inline comment added at `tools/security_tools.py:438` explaining the decorator-time constraint | ✅ Resolved — splitting would have been API churn; comment is the right tradeoff |
| Suggestion 1: hardcoded bulk action error message | `VALID_BULK_ACTIONS` switched to tuple (stable order), `_BULK_ACTIONS_STR` mirrors the `_FILE_ACL_ACTIONS_STR` pattern | ✅ Clean |
| Suggestion 2: empty-string clear semantics | Docstrings on `update_command_acl` and `update_file_acl` now spell out "pass '' to clear, omit to leave unchanged" — landed on the right side (update, where clearing actually applies) | ✅ Better placement than my original suggestion |
| Suggestion 3: exactly-100 boundary test | `test_bulk_exactly_100_server_ids_succeeds` added at `tests/test_security_tools.py:721`; verifies success + payload shape + endpoint routing for the boundary | ✅ Good test |
| Suggestion 4: manual cross-type rebind verification | Author noted this is blocked on shipping (dev MCP server doesn't expose `update_server_acl` until merge) | ⚠️ Reasonable — flagging for the merger below |

### New observations (0 critical, 0 warnings, 0 suggestions)

No new issues found. The `update_server_acl` cross-type rebind logic at `tools/security_tools.py:381-386` (nulling the opposing token field on PATCH) is the only server-coupled behavior that unit tests can't fully verify. The WHY-comment at line 379-380 is clear about the server's `exactly_one_token_type` constraint.

### Pre-merge note (for whoever merges)

- [ ] **Manual cross-type token rebind verification against a dev workspace.** Verify that `update_server_acl` with an api-token-bound ACL → service token (and vice versa) actually succeeds against the live API, not just at the request-shape level. If this surfaces a misread of the server check constraint, it would show up as a confusing `ACL_EXISTS` in production. The author has flagged this as the test plan's only open item — it's not blocking the code review, but worth closing before tagging a release that depends on this tool.

### Highlights of this iteration

- Followed the existing `_FILE_ACL_ACTIONS_STR` pattern for the new `_BULK_ACTIONS_STR` constant — consistent code shape, no drift risk if `VALID_BULK_ACTIONS` ever grows
- Docstring clarification on `update_*` (not `create_*`) is more precise than my original suggestion — clearing is an update-side concern; placing the note on create would have been misleading
- One-line decorator comment is exactly enough — explains the constraint without bloating the code

---

**Verdict**: APPROVED — all prior feedback addressed, no new issues. Manual rebind verification remains the merger's checklist item.

**Checklist**:
- [x] All prior review items resolved
- [x] No new code-review issues introduced
- [x] Tests cover the boundary case (100-item bulk cap)
- [x] Stack conventions followed (`webftp_tools`/`iam_tools` pattern)
- [ ] Cross-type token rebind verified against dev workspace (still author's open item)

---
🤖 Auto-review by junho-bot during junho226's leave. If this review looks off, please disregard — junho will follow up upon return.

**Review by eunyoung14 — APPROVED** [review]
